### PR TITLE
 [GEOS-10058] Welcome Page Layout

### PR DESCRIPTION
**Pull request by jodygarnett**
[GEOS-10058](https://osgeo-org.atlassian.net/browse/GEOS-10058)

See https://github.com/geoserver/geoserver/wiki/GSIP-202 proposal for details

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by jodygarnett** [comment]
Sharing draft pull request for feedback.

Notes:

* Reviewed the the existing ``CapabilitiesHomePageLinkProvider`` which was setup to return a component; but in reality the component was always a ``CapabilitiesHomePagePanel`` which does not have enough visibility to see ServiceInfo objects.  Decided to setup a ``ServiceDescriptionProvider`` extension to return a data structure (ServiceDescriptino and ServiceLinkDescription)  which can be processed for display.  We can deprecate ``CapabilitiesHomePageLinkProvider`` when transition is complete.

* Experimenting with using page properties to include workspace and layer information in URL (``ne``, ``ne/roads``, ``?workspace=ne``, ``?workspace=ne&layer=roads``) so we can share a URL to a workspace's  services. The logic for how this is handled in dispatcher is spread out in a number of locations.

File: src/web/core/src/main/java/org/geoserver/web/CapabilitiesHomePagePanel.java [hunk 1]
```diff
@@ -34,15 +34,19 @@ public class CapabilitiesHomePagePanel extends Panel {
...
         String service;
 
+        /** Service version */
         Version version;
 
+        /** GetCapabilities service description */
         String capsLink;
 
-        public CapsInfo(String service, Version version, String capsLink) {
+        public CapsInfo(String service, Version version, String capsLink){
```
Comment (bradh) [inline src/web/core/src/main/java/org/geoserver/web/CapabilitiesHomePagePanel.java:49]: I guess this is going to get reformatted.

File: src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java [hunk 2]
```diff
@@ -42,46 +45,77 @@
...
 /**
- * Home page, shows just the introduction and the capabilities link
+ * Home page, shows introduction for each kind of service along with any service links.
  *
- * <p>This page uses the {@link CapabilitiesHomePageLinkProvider} extension point to enable other
- * modules to contribute links for GetCapabilities documents. The default {@link
- * ServiceInfoCapabilitiesProvider} contributes the capabilities links for all the available {@link
- * ServiceInfo} implementations. Other extension point implementations may contribute service
- * description document links non backed by ServiceInfo objects.
+ * <p>This page uses {@link ServiceDescriptionProvider} extension point allow other modules describe
```
Comment (bradh) [inline src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java:50]:
```suggestion
 * <p>This page uses the {@link ServiceDescriptionProvider} extension point to allow other modules to describe
```

File: src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java [hunk 3]
```diff
@@ -42,46 +45,77 @@
...
+ * web services, and the web service links.
+ *
+ * <p>The {@link CapabilitiesHomePageLinkProvider} extension point to enable other modules to
```
Comment (bradh) [inline src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java:53]:
```suggestion
 * <p>The {@link CapabilitiesHomePageLinkProvider} extension point enables other modules to
```

File: src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java [hunk 4]
```diff
@@ -42,46 +45,77 @@
...
+ * contribute components. The default {@link ServiceInfoCapabilitiesProvider} contributes the
+ * capabilities links for all the available {@link ServiceInfo} implementations that were not
+ * covered by ServiceDescriptionProvider. Other extension point implementations may contribute
```
Comment (bradh) [inline src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java:56]:
```suggestion
 * covered by the ServiceDescriptionProvider. Other extension point implementations may contribute
```

File: src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java [hunk 5]
```diff
@@ -42,46 +45,77 @@
...
+ * service description document links not backed by ServiceInfo objects.
+ *
+ * <p>The {@link GeoServerHomePageContentProvider} is used by modules to contribute information, status
+ * and warnings.</p>
+ *
+ * <p>
+ * The page has built-in functionality providing administrators with a configuration summary.
+ * </p>
+ *
+ * <p>This page can change between gloabl-service, workspace service and layer service.
```
Comment (bradh) [inline src/web/core/src/main/java/org/geoserver/web/GeoServerHomePage.java:66]:
```suggestion
 * <p>This page can change between global service, workspace service and layer service.
```

File: src/web/core/src/main/java/org/geoserver/web/ServiceInfoCapabilitiesProvider.java [hunk 6]
```diff
@@ -22,6 +24,71 @@
  * @see CapabilitiesHomePagePanel
  */
 public class ServiceInfoCapabilitiesProvider implements CapabilitiesHomePageLinkProvider {
+    /*
```
Comment (bradh) [inline src/web/core/src/main/java/org/geoserver/web/ServiceInfoCapabilitiesProvider.java:27]: Assume this is associated with draft status.

File: src/web/core/src/main/java/org/geoserver/web/ServicesPanel.java [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+     */
+    public static class ServiceDescription implements Serializable, Comparable<ServiceDescription> {
+        private static final long serialVersionUID = -7406652617944177247L;
+
+        /**
+         * Workspace prefix for virtual web service, may be null for global services.
+         *
+         * Forced to lowercase for ease of comparison.
+         */
+        final String workspace;
```
Comment (bradh) [inline src/web/core/src/main/java/org/geoserver/web/ServicesPanel.java:54]: You have (at least) getters, so these probably would be better `private`.

**Review by bradh — COMMENTED** [review]
Some mixed thoughts.

**Comment by aaime** [comment]
I was thinking about the last iteration of the home page I've seen yesterday, it looks like it's veering away from the other GeoServer pages in terms of colors, font sizes, shape of the buttons and the like, more than what it was [already proposed](https://github.com/geoserver/geoserver/wiki/GSIP-202#initial-design) (colors and font sizes were already a bit off, but the utility buttons at the right somehow managed to bring back some of the feeling).

GeoServer has an active branding guide, voted as part of GSIP-26, which you can find here:
https://github.com/geoserver/geoserver/wiki/GeoServer-Branding
and in particular in this PDF document:
https://github.com/geoserver/geoserver/wiki/GeoServer_BrandingGuide.pdf

Could you please make sure the overall home page is a clear implementation of the guidelines? Visually, it should not be hard to check, login as admin, and verify that switching between the home page and inner pages is a seamless visual experience.

**Comment by jodygarnett** [comment]
The branding guide is about logo use and is unaffected…

The font sizes and so on are a mess (too big too bold visually) which Michel is going help clean up with his css experience.

**Comment by aaime** [comment]
> The branding guide is about logo use and is unaffected…

I think you're making a read of it that's too verbatim. The GeoServer pages have a consistent choice of colors and graphics that form (most of the time) a consistent look, inspired from the logo. The home page of GeoServer needs to reflect that, be consistent.

File: src/web/core/src/main/java/org/geoserver/web/CapabilitiesHomePagePanel.java (hunk 1 above)
Comment (jodygarnett) [reply src/web/core/src/main/java/org/geoserver/web/CapabilitiesHomePagePanel.java]: yes it was still to be formatted

**Comment by jodygarnett** [comment]
Moved to https://github.com/geoserver/geoserver/pull/6122 to track [welcome](https://github.com/geoserver/geoserver/tree/welcome) branch.

**Pull request closed without merging**